naturalcrit/homebrewery
# Rework combobox.jsx to functional component

## Description

This is a rework of the Combobox component.  It is now a Functional component.  It makes a few small changes to structure (the options are all buttons now).  It also adds better keyboard navigation (which is pretty much all of the "new lines of code").  Beyond that, it is mostly the same at this point.

Right now, only the Language picker uses this component.  I started to rework it because I want to use it in the TagInput component (aka StringArrayEditor).  

Beyond whatever is suggested in review, some additional things I want to do:

- [x] add a "expand"/"open" button to the right side of the input as a button to open it (in truth, the whole input toggles the menu so this will just be a little visual indicator).
- [ ] a quick pass on accessibility issues (roles, hidden navigation)
- [ ] get the validators working with it again.

Then, as a subsequent PR, I want to work out a good way to get submenus to work.


## Related Issues or Discussions

- Related #3735

## QA Instructions, Screenshots, Recordings

Try out the Language picker, using your mouse and navigating it with the keyboard.  You should be able to:

- Tab into it and focus the input
- Space to open the menu with all options, Space again to close it
- Type any characters to open menu to filtered options
- Typing Space should still work to add whitespace, without opening/closing the menu, as long as the input isn't otherwise empty
- Esc to close the drop down
- Enter to select an option
- ArrowDown from input to open and navigate through menu, cycling to first option if reached the bottom
- ArrowUp to move through menu, focusing the input when you move past the first option
- when focus is on menu option, be able to use Backspace to move focus to input and delete characters.
- Tab out to next input closes the dropdown and moves focus.
- entering a value that filters to zero results displays a 'no matches' item.

Possibly some of these behaviors will change as i look at WCAG for comboboxes and I'll try to update this list.  

### Reviewer Checklist

*Reviewers, refer to this list when testing features, or suggest new items *

<details><summary>Copy this list</summary>

```
	- [ ] Tab into it and focus the input
	- [ ] Space to open the menu with all options, Space again to close it
	- [ ] Type any characters to open menu to filtered options
	- [ ] Typing Space should still work to add whitespace, without opening/closing the menu, as long as the input isn't otherwise empty
	- [ ] Esc to close the drop down
	- [ ] Enter to select an option
	- [ ] ArrowDown from input to open and navigate through menu, cycling to first option if reached the bottom
	- [ ] ArrowUp to move through menu, focusing the input when you move past the first option
	- [ ] when focus is on menu option, be able to use Backspace to move focus to input and delete characters.
	- [ ] Tab out to next input closes the dropdown and moves focus.
	- [ ] entering a value that filters to zero results displays a 'no matches' item.
  - [ ] Still correctly sets the `lang` attribute and is preserved with the brew metadata
```

</details>
